tests/codegen-units/partitioning/auxiliary/cgu_generic_function.rs

```diff
@@ -11,10 +11,21 @@ pub fn foo<T>(x: T) -> (T, u32, i8) {
 #[inline(never)]
 fn bar<T>(x: T) -> (T, Struct) {
     let _ = not_exported_and_not_generic(0);
+    exported_and_generic::<u32>(0);
     (x, Struct(1))
 }
 
+pub static F: fn(u32) -> u32 = exported_and_generic::<u32>;
+
 // These should not contribute to the codegen items of other crates.
+
+// This is generic, but it's only instantiated with a u32 argument and that instantiation is present
+// in the local crate (see F above).
```

Comment:
Is this testing what we want? I was expecting to see a cross-crate call to a `#[inline(never)]` generic function in a test, because I think the point of this PR is to change the behavior for such calls, right?

Reply:
I think this is indeed checking that behavior, though it's rather obscure. As per the comment just above ("These should not contribute...") we are implicitly checking that these functions are absent in the mono-items of the downstream crate ([tests/codegen-units/partitioning/extern-generic.rs](https://github.com/rust-lang/rust/blob/master/tests/codegen-units/partitioning/extern-generic.rs)) since they're not listed as `MONO-ITEM` declarations in that file. They *are* referenced though via the `foo` function above which has to get codegen'd in the downstream crate since it is generic and never called in this one.

I suppose we could try to write a more direct test case? But this also tests that the property holds transitively (i.e., we don't codegen the called function twice with inline(never) even if it's not directly being called).

IIRC, before my changes, this would fail.

Reply:
Oh I see! This is very interconnected, and I should really be used to that by now.

Reply:
Also I checked and yes the test does fail without the rest of this PR.